README.rst
The **ibm2ieee** package provides NumPy universal functions ("ufuncs") for
converting IBM single-precision and double-precision hexadecimal floats
to the IEEE 754-format floats used by Python and NumPy on almost all
current platforms.


Features
--------

- Fast: 200-400 million values converted per second on a typical modern
  machine, assuming normal inputs.
- Correct: converted results are correctly rounded, according to the default
  IEEE 754 round-ties-to-even rounding mode. Corner cases (overflow, underflow,
  subnormal results, signed zeros, non-normalised input) are all handled
  correctly. Where the rounded converted value is out of range for the target
  type, an appropriately-signed infinity is returned.
- Handles both single-precision and double-precision input and output formats.

Portability note: the conversion functions provided in this module assume that
``numpy.float32`` and ``numpy.float64`` are based on the standard IEEE 754
binary32 and binary64 floating-point formats. This is true on the overwhelming
majority of current platforms, but is not guaranteed by the relevant language
standards.


Usage
-----

The package provides two functions:

- ``ibm2float32`` converts IBM single- or double-precision data to
  IEEE 754 single-precision values, in ``numpy.float32`` format.

- ``ibm2float64`` converts IBM single- or double-precision data to
  IEEE 754 double-precision values, in ``numpy.float64`` format.

For both functions, IBM single-precision input data must be represented
using the ``numpy.uint32`` dtype, while IBM double-precision inputs must
be represented using ``numpy.uint64``.

Both functions assume that the IBM data have been converted to NumPy integer
format in such a way that the most significant bits of the floating-point
number become the most significant bits of the integer values. So when decoding
byte data representing IBM hexadecimal floating-point numbers, it's important
to take the endianness of the byte data into account. See the Examples section
below for an example of converting big-endian byte data.


Examples
--------

>>> import numpy
>>> from ibm2ieee import ibm2float32, ibm2float64
>>> ibm2float32(numpy.uint32(0xc1180000))
-1.5
>>> type(ibm2float32(numpy.uint32(0xc1180000)))
<class 'numpy.float32'>
>>> ibm2float32(numpy.uint64(0x413243f6a8885a31))
3.1415927
>>> ibm2float32(numpy.uint32(0x61100000))
inf
>>> ibm2float64(numpy.uint32(0xc1180000))
-1.5
>>> ibm2float64(numpy.uint64(0x413243f6a8885a31))
3.141592653589793
>>> ibm2float64(numpy.uint32(0x61100000))
3.402823669209385e+38
>>> input_array = numpy.arange(
        0x40fffffe, 0x41000002, dtype=numpy.uint32).reshape(2, 2)
>>> input_array
array([[1090519038, 1090519039],
       [1090519040, 1090519041]], dtype=uint32)
>>> ibm2float64(input_array)
array([[9.99999881e-01, 9.99999940e-01],
       [0.00000000e+00, 9.53674316e-07]])

When converting byte data read from a file, it's important to know the
endianness of that data (which is frequently big-endian in historical data
files using IBM hex floating-point). Here's an example of converting IBM
single-precision data stored in big-endian form to ``numpy.float32``. Note the
use of the ``'>u4'`` dtype when converting the bytestring to a NumPy ``uint32``
array. For little-endian input data, you would use ``'<u4'`` instead.

>>> input_data = b'\xc12C\xf7\xc1\x19!\xfb\x00\x00\x00\x00A\x19!\xfbA2C\xf7'
>>> input_as_uint32 = numpy.frombuffer(input_data, dtype='>u4')
>>> input_as_uint32
array([3241296887, 3239649787,          0, 1092166139, 1093813239],
      dtype=uint32)
>>> ibm2float32(input_as_uint32)
array([-3.141593, -1.570796,  0.      ,  1.570796,  3.141593],
      dtype=float32)


Notes on the formats
--------------------

The IBM single-precision format has a precision of 6 hexadecimal digits, which
in practice translates to a precision of 21-24 bits, depending on the binade
that the relevant value belongs to. IEEE 754 single-precision has a precision
of 24 bits. So all not-too-small, not-too-large IBM single-precision values can
be translated to IEEE 754 single-precision values with no loss of precision.
However, the IBM single precision range is larger than the corresponding IEEE
754 range, so extreme IBM single-precision values may overflow to infinity,
underflow to zero, or be rounded to a subnormal value when converted to IEEE
754 single-precision.

For double-precision conversions, the tradeoff works the other way: the IBM
double-precision format has an effective precision of 53-56 bits, while IEEE
754 double-precision has 53-bit precision. So most IBM values will be rounded
when converted to IEEE 754. However, the IEEE 754 double-precision range is
larger than that of IBM double-precision, so there's no danger of overflow,
underflow, or reduced-precision subnormal results when converting IBM
double-precision to IEEE 754 double-precision.

Every IBM single-precision value can be exactly represented in IEEE 754
double-precision, so if you want a lossless representation of IBM
single-precision data, use ``ibm2float64``.

Note that the IBM formats do not allow representations of special values like
infinities and NaNs. However, signed zeros are representable, and the sign of a
zero is preserved under all conversions.


Installation
------------

The latest release of ibm2ieee is available from the Python Package Index, at
https://pypi.org/project/ibm2ieee. It can be installed with ``pip`` in the
usual way::

    pip install ibm2ieee

Wheels are provided for common platforms and Python versions. If installing
from source, note that ibm2ieee includes a C extension, so you'll need the
appropriate compiler on your system to be able to install.

ibm2ieee requires Python >= 3.7.


License
-------

(C) Copyright 2018-2023 Enthought, Inc., Austin, TX
All rights reserved.

This software is provided without warranty under the terms of the BSD
license included in LICENSE.txt and may be redistributed only under
the conditions described in the aforementioned license. The license
is also available online at http://www.enthought.com/licenses/BSD.txt

Thanks for using Enthought open source!


CHANGES
Changelog for ibm2ieee
======================

Release 1.3.3
-------------

Release date: 2023-10-08

This bugfix release fixes the wheel building configuration by removing
support for cp312-manylinux_i686.

Release 1.3.2
-------------

Release date: 2023-10-08

This bugfix release extends the wheel building configuration to support
Python 3.12.

Release 1.3.1
-------------

Release date: 2023-03-31

This bugfix release fixes the wheel building configuration. In the process,
we also:

- Move most of the package configuration to `pyproject.toml`
- Drop the `ibm2ieee.version` module and the `ibm2ieee.__version__` attribute.

Release 1.3.0
-------------

Release date: 2023-03-31

This minor release updates supported Python versions. There are no changes to
the functional core.

- Python 3.6 is no longer supported; the minimum required version is Python
  3.7.

Release 1.2.0
-------------

Release date: 2022-08-16

This minor release updates supported Python versions and various parts of the
continuous integration and development workflow. There are no changes to the
functional core.

- Apply black and isort to the code, and enforce style in CI checks. (#33)
- Update PyPI wheel-build-and-upload workflow to use cibuildwheel. (#31)
- Update .gitignore. (#29)
- Use subTest in repetitive tests. (#27)
- Update build infrastructure; use ``oldest-supported-numpy`` in build
  requirements; extend to Python 3.10 and 3.11. (#30)
- Update copyright end year. (#28)
- Update installation notes to mention wheels. (#24)
- Add workflow to test PyPI wheels. (#23, #32, #34)


Release 1.1.0
-------------

Release date: 2021-01-10

This minor release updates the supported Python versions, and makes
various distributional and developer-facing changes. Python 2.7
and Python 3.5 are no longer supported, and Python 3.9 is now
explicitly supported. There are no changes to the functional core.

- Move to using GitHub Actions for PR tests. (#22)

- Add workflow to automatically build and publish wheels for
   releases. (#20)

- Update copyright header end year; update copyright header form to match the
  official Enthought style used elsewhere in the Enthought Tool Suite, and
  add a flake8 check for correct and up-to-date copyright headers. (#19)

- Drop support for Python 2.7 and Python 3.5. The package now requires
  Python >= 3.6. (#18)

- Add explicit support for Python 3.9. (#17)


Release 1.0.2
-------------

Release date: 2020-02-09

This is a bugfix release, addressing distribution and development
issues along with minor cleanups to the core.

- Require NumPy >= 1.14.5 at runtime; build against NumPy == 1.14.5
  at installation time. (PR #8)
- Replace the ``packaging`` dependency of the test suite with
  ``setuptools``; add explicit dependency of the test suite on ``six``.
  (PR #11)
- Add classifiers for Python 3.7 and 3.8, and add Python 3.7 and 3.8
  to the test matrix for Travis and Appveyor. (PR #8)
- Silence some current and future compiler and linter warnings in
  the core C code. (PR #15)
- Apply a micro-optimization for the 64-bit IBM to 32-bit IEEE conversion,
  replacing a 64-bit addition with a 32-bit addition. (PR #15)


Release 1.0.1
-------------

Release date: 2018-12-04

Bugfix release, fixing another distribution issue.

- Add ``pyproject.toml`` file encapsulating build requirements. With this, a
  ``pip install ibm2ieee`` should automatically download NumPy before
  running the setup script. (PR #7)


Release 1.0.0
-------------

Release date: 2018-12-04

First real release. Fixes some distribution issues with the 0.1.0 release. Code is
otherwise unchanged.

- Fix misuse of ``unicode_literals`` in the setup.py file.
- Fix trailing newline in short description, which caused the README not
  to render properly on PyPI.
- Boldly bump the version number to 1.0.0, and the Development Status classifier
  to Production/Stable.


Release 0.1.0
-------------

Release date: 2018-12-04

Initial release uploaded to PyPI.


ibm2ieee/test/__init__.py
# (C) Copyright 2018-2023 Enthought, Inc., Austin, TX
# All rights reserved.
#
# This software is provided without warranty under the terms of the BSD
# license included in LICENSE.txt and may be redistributed only under
# the conditions described in the aforementioned license. The license
# is also available online at http://www.enthought.com/licenses/BSD.txt
#
# Thanks for using Enthought open source!


ibm2ieee/test/test_version.py
# (C) Copyright 2018-2023 Enthought, Inc., Austin, TX
# All rights reserved.
#
# This software is provided without warranty under the terms of the BSD
# license included in LICENSE.txt and may be redistributed only under
# the conditions described in the aforementioned license. The license
# is also available online at http://www.enthought.com/licenses/BSD.txt
#
# Thanks for using Enthought open source!

import unittest

import pkg_resources

import ibm2ieee.version


class TestVersion(unittest.TestCase):
    def test_version_string(self):
        version = ibm2ieee.version.version
        self.assertIsInstance(version, str)

        # Check that version number is normalised and complies with PEP 440.
        version_object = pkg_resources.parse_version(version)
        self.assertEqual(str(version_object), version)

    def test_top_level_package_version(self):
        self.assertEqual(
            ibm2ieee.__version__,
            ibm2ieee.version.version,
        )


ibm2ieee/_ibm2ieee.c
// (C) Copyright 2018-2023 Enthought, Inc., Austin, TX
// All rights reserved.
//
// This software is provided without warranty under the terms of the BSD
// license included in LICENSE.txt and may be redistributed only under
// the conditions described in the aforementioned license. The license
// is also available online at http://www.enthought.com/licenses/BSD.txt
//
// Thanks for using Enthought open source!

#include "Python.h"
#include "numpy/ndarraytypes.h"
#include "numpy/ufuncobject.h"


/* Format-related masks */

#define IBM32_SIGN ((npy_uint32)0x80000000U)
#define IBM32_EXPT ((npy_uint32)0x7f000000U)
#define IBM32_FRAC ((npy_uint32)0x00ffffffU)
#define IBM32_TOP  ((npy_uint32)0x00f00000U)
#define TIES_TO_EVEN_MASK32 ((npy_uint32)0xfffffffdU)

#define IBM64_SIGN ((npy_uint64)0x8000000000000000U)
#define IBM64_EXPT ((npy_uint64)0x7f00000000000000U)
#define IBM64_FRAC ((npy_uint64)0x00ffffffffffffffU)
#define IBM64_TOP  ((npy_uint64)0x00f0000000000000U)
#define TIES_TO_EVEN_MASK64 ((npy_uint64)0xfffffffffffffffdU)

/* Masks used for 3-bit and 32-bit rounded right-shifts of a 64-bit quantity.
   The masks comprise the parity bit and the trailing bits for the shift. */
#define TIES_TO_EVEN_RSHIFT3  ((npy_uint64)0x000000000000000bU)
#define TIES_TO_EVEN_RSHIFT32 ((npy_uint64)0x000000017fffffffU)

#define IEEE32_MAXEXP 254     /* Maximum biased exponent for finite values. */
#define IEEE32_INFINITY ((npy_uint32)0x7f800000U)

/* Constant used to count number of leading bits in a nonzero hex digit
   via `(BITCOUNT_MAGIC >> (hex_digit*2)) & 3U`. */
#define BITCOUNT_MAGIC ((npy_uint32)0x000055afU)


/* IBM single-precision bit pattern to IEEE single-precision bit pattern. */

static npy_uint32
ibm32ieee32(npy_uint32 ibm)
{
    /* Overflow and underflow possible; rounding can only happen
       in subnormal cases. */
    int ibm_expt, ieee_expt, leading_zeros;
    npy_uint32 ibm_frac, top_digit;
    npy_uint32 ieee_sign, ieee_frac;

    ieee_sign = ibm & IBM32_SIGN;
    ibm_frac = ibm & IBM32_FRAC;

    /* Quick return for zeros. */
    if (!ibm_frac) {
        return ieee_sign;
    }

    /* Reduce shift by 2 to get a binary exponent from the hex exponent. */
    ibm_expt = (int)((ibm & IBM32_EXPT) >> 22);

    /* Normalise significand, then count leading zeros in top hex digit. */
    top_digit = ibm_frac & IBM32_TOP;
    while (top_digit == 0) {
        ibm_frac <<= 4;
        ibm_expt -= 4;
        top_digit = ibm_frac & IBM32_TOP;
    }
    leading_zeros = (int)((BITCOUNT_MAGIC >> (top_digit >> 19)) & 3U);
    ibm_frac <<= leading_zeros;

    /* Adjust exponents for the differing biases of the formats: the IBM bias
       is 64 hex digits, or 256 bits. The IEEE bias is 127. The difference is
       -129; we get an extra -1 from the different significand representations
       (0.f for IBM versus 1.f for IEEE), and another -1 to compensate for an
       evil trick that saves an operation on the fast path: we don't remove the
       hidden 1-bit from the IEEE significand, so in the final addition that
       extra bit ends in incrementing the exponent by one. */
    ieee_expt = ibm_expt - 131 - leading_zeros;

    if (ieee_expt >= 0 && ieee_expt < IEEE32_MAXEXP) {
        /* normal case; no shift needed */
        ieee_frac = ibm_frac;
        return ieee_sign + ((npy_uint32)ieee_expt << 23) + ieee_frac;
    }
    else if (ieee_expt >= IEEE32_MAXEXP) {
        /* overflow */
        return ieee_sign + IEEE32_INFINITY;
    }
    else if (ieee_expt >= -32) {
        /* possible subnormal result; shift significand right by -ieee_expt
           bits, rounding the result with round-ties-to-even.

           The round-ties-to-even code deserves some explanation: out of the
           bits we're shifting out, let's call the most significant bit the
           "rounding bit", and the rest the "trailing bits". We'll call the
           least significant bit that *isn't* shifted out the "parity bit".
           So for an example 5-bit shift right, we'd label the bits as follows:

           Before the shift:

                   ...xxxprtttt
                              ^
              msb            lsb

           After the shift:

                        ...xxxp
                              ^
              msb            lsb

           with the result possibly incremented by one.

           For round-ties-to-even, we need to round up if both (a) the rounding
           bit is 1, and (b) either the parity bit is 1, or at least one of the
           trailing bits is 1. We construct a mask that has 1-bits in the
           parity bit position and trailing bit positions, and use that to
           check condition (b). So for example in the 5-bit shift right, the
           mask looks like this:

                   ...000101111 : mask
                   ...xxxprtttt : ibm_frac
                              ^
              msb            lsb

           We then shift right by (shift - 1), add 1 if (ibm & mask) is
           nonzero, and then do a final shift by one to get the rounded
           value. Note that this approach avoids the possibility of
           trying to shift a width-32 value by 32, which would give
           undefined behaviour (see C99 6.5.7p3).
         */
        npy_uint32 mask = ~(TIES_TO_EVEN_MASK32 << (-1 - ieee_expt));
        npy_uint32 round_up = (ibm_frac & mask) > 0U;
        ieee_frac = ((ibm_frac >> (-1 - ieee_expt)) + round_up) >> 1;
        return ieee_sign + ieee_frac;
    }
    else {
        /* underflow to zero */
        return ieee_sign;
    }
}


/* IBM double-precision bit pattern to IEEE single-precision bit pattern. */

static npy_uint32
ibm64ieee32(npy_uint64 ibm)
{
    /* Overflow and underflow possible; rounding can occur in both
       normal and subnormal cases. */
    int ibm_expt, ieee_expt, leading_zeros;
    npy_uint64 ibm_frac, top_digit;
    npy_uint32 ieee_sign, ieee_frac;

    ieee_sign = (ibm & IBM64_SIGN) >> 32;
    ibm_frac = ibm & IBM64_FRAC;

    /* Quick return for zeros. */
    if (!ibm_frac) {
        return ieee_sign;
    }

    /* Reduce shift by 2 to get a binary exponent from the hex exponent. */
    ibm_expt = (int)((ibm & IBM64_EXPT) >> 54);

    /* Normalise significand, then count leading zeros in top hex digit. */
    top_digit = ibm_frac & IBM64_TOP;
    while (top_digit == 0) {
        ibm_frac <<= 4;
        ibm_expt -= 4;
        top_digit = ibm_frac & IBM64_TOP;
    }
    leading_zeros = (int)((BITCOUNT_MAGIC >> (top_digit >> 51)) & 3U);

    ibm_frac <<= leading_zeros;
    ieee_expt = ibm_expt - 131 - leading_zeros;

    if (ieee_expt >= 0 && ieee_expt < IEEE32_MAXEXP) {
        /* normal case; shift right 32, with round-ties-to-even */
        npy_uint32 round_up = (ibm_frac & TIES_TO_EVEN_RSHIFT32) > 0U;
        ieee_frac = ((npy_uint32)(ibm_frac >> 31) + round_up) >> 1;
        return ieee_sign + ((npy_uint32)ieee_expt << 23) + ieee_frac;
    }
    else if (ieee_expt >= IEEE32_MAXEXP) {
        /* overflow */
        return ieee_sign + IEEE32_INFINITY;
    }
    else if (ieee_expt >= -32) {
        /* possible subnormal; shift right with round-ties-to-even */
        npy_uint64 mask = ~(TIES_TO_EVEN_MASK64 << (31 - ieee_expt));
        npy_uint32 round_up = (ibm_frac & mask) > 0U;
        ieee_frac = (
            (npy_uint32)(ibm_frac >> (31 - ieee_expt)) + round_up) >> 1;
        return ieee_sign + ieee_frac;
    }
    else {
        /* underflow to zero */
        return ieee_sign;
    }
}


/* IBM single-precision bit pattern to IEEE double-precision bit pattern. */

static npy_uint64
ibm32ieee64(npy_uint32 ibm)
{
    /* This is the simplest of the four cases: there's no need to check for
       overflow or underflow, no possibility of subnormal output, and never
       any rounding. */
    int ibm_expt, ieee_expt, leading_zeros;
    npy_uint32 ibm_frac, top_digit;
    npy_uint64 ieee_sign, ieee_frac;

    ieee_sign = (npy_uint64)(ibm & IBM32_SIGN) << 32;
    ibm_frac = ibm & IBM32_FRAC;

    /* Quick return for zeros. */
    if (!ibm_frac) {
        return ieee_sign;
    }

    /* Reduce shift by 2 to get a binary exponent from the hex exponent. */
    ibm_expt = (int)((ibm & IBM32_EXPT) >> 22);

    /* Normalise significand, then count leading zeros in top hex digit. */
    top_digit = ibm_frac & IBM32_TOP;
    while (top_digit == 0) {
        ibm_frac <<= 4;
        ibm_expt -= 4;
        top_digit = ibm_frac & IBM32_TOP;
    }
    leading_zeros = (int)((BITCOUNT_MAGIC >> (top_digit >> 19)) & 3U);

    /* Adjust exponents for the differing biases of the formats: the IBM bias
       is 64 hex digits, or 256 bits. The IEEE bias is 1023. The difference is
       767; we get an extra -1 from the different significand representations
       (0.f for IBM versus 1.f for IEEE), and another -1 to compensate for an
       evil trick that saves an operation: we don't remove the hidden 1-bit
       from the IEEE significand, so in the final addition that extra bit ends
       in incrementing the exponent by one. */
    ieee_expt = ibm_expt + 765 - leading_zeros;
    ieee_frac = (npy_uint64)ibm_frac << (29 + leading_zeros);
    return ieee_sign + ((npy_uint64)ieee_expt << 52) + ieee_frac;
}


/* IBM double-precision bit pattern to IEEE double-precision bit pattern. */

static npy_uint64
ibm64ieee64(npy_uint64 ibm)
{
    /* No overflow or underflow possible, but the precision of the
       IBM double-precision format exceeds that of its IEEE counterpart,
       so we'll frequently need to round. */
    int ibm_expt, ieee_expt, leading_zeros;
    npy_uint64 ibm_frac, top_digit;
    npy_uint64 ieee_sign, ieee_frac, round_up;

    ieee_sign = ibm & IBM64_SIGN;
    ibm_frac = ibm & IBM64_FRAC;

    /* Quick return for zeros. */
    if (!ibm_frac) {
        return ieee_sign;
    }

    /* Reduce shift by 2 to get a binary exponent from the hex exponent. */
    ibm_expt = (int)((ibm & IBM64_EXPT) >> 54);

    /* Normalise significand, then count leading zeros in top hex digit. */
    top_digit = ibm_frac & IBM64_TOP;
    while (top_digit == 0) {
        ibm_frac <<= 4;
        ibm_expt -= 4;
        top_digit = ibm_frac & IBM64_TOP;
    }
    leading_zeros = (int)((BITCOUNT_MAGIC >> (top_digit >> 51)) & 3U);

    ibm_frac <<= leading_zeros;
    ieee_expt = ibm_expt + 765 - leading_zeros;

    /* Right-shift by 3 bits (the difference between the IBM and IEEE
       significand lengths), rounding with round-ties-to-even. */
    round_up = (ibm_frac & TIES_TO_EVEN_RSHIFT3) > 0U;
    ieee_frac = ((ibm_frac >> 2) + round_up) >> 1;
    return ieee_sign + ((npy_uint64)ieee_expt << 52) + ieee_frac;
}

/* NumPy ufunc wrapper for ibm32ieee32 */

static void
ibm32ieee32_ufunc(
    char **args, npy_intp const *dimensions, npy_intp const *steps, void *data)
{
    npy_intp i;
    npy_intp n = dimensions[0];
    char *in = args[0], *out = args[1];
    npy_intp in_step = steps[0], out_step = steps[1];

    for (i = 0; i < n; i++) {
        *((npy_uint32 *)out) = ibm32ieee32(*(npy_uint32 *)in);
        in += in_step;
        out += out_step;
    }
}

/* NumPy ufunc wrapper for ibm64ieee32 */

static void
ibm64ieee32_ufunc(
    char **args, npy_intp const *dimensions, npy_intp const *steps, void *data)
{
    npy_intp i;
    npy_intp n = dimensions[0];
    char *in = args[0], *out = args[1];
    npy_intp in_step = steps[0], out_step = steps[1];

    for (i = 0; i < n; i++) {
        *((npy_uint32 *)out) = ibm64ieee32(*(npy_uint64 *)in);
        in += in_step;
        out += out_step;
    }
}

/* NumPy ufunc wrapper for ibm32ieee64 */

static void
ibm32ieee64_ufunc(
    char **args, npy_intp const *dimensions, npy_intp const *steps, void *data)
{
    npy_intp i;
    npy_intp n = dimensions[0];
    char *in = args[0], *out = args[1];
    npy_intp in_step = steps[0], out_step = steps[1];

    for (i = 0; i < n; i++) {
        *((npy_uint64 *)out) = ibm32ieee64(*(npy_uint32 *)in);
        in += in_step;
        out += out_step;
    }
}

/* NumPy ufunc wrapper for ibm64ieee64 */

static void
ibm64ieee64_ufunc(
    char **args, npy_intp const *dimensions, npy_intp const *steps, void *data)
{
    npy_intp i;
    npy_intp n = dimensions[0];
    char *in = args[0], *out = args[1];
    npy_intp in_step = steps[0], out_step = steps[1];

    for (i = 0; i < n; i++) {
        *((npy_uint64 *)out) = ibm64ieee64(*(npy_uint64 *)in);
        in += in_step;
        out += out_step;
    }
}

PyDoc_STRVAR(ibm2float32_docstring,
             "\
Convert IBM-format single- or double-precision float data, represented\n\
using types np.uint32 or np.uint64 (respectively), to np.float32.\n\
\n\
Examples\n\
--------\n\
>>> ibm2float32(np.uint32(0xc1180000))\n\
-1.5\n\
>>> ibm2float32(np.uint64(0x413243f6a8885a31))\n\
3.1415927\n\
>>> ibm2float32(np.uint32(0x61100000))\n\
inf\n\
");

PyUFuncGenericFunction ibm2float32_funcs[2] = {
    /* Cast required to suppress compiler warning for NumPy < 1.19.0.
       See https://github.com/numpy/numpy/issues/15252. */
    (PyUFuncGenericFunction)&ibm32ieee32_ufunc,
    (PyUFuncGenericFunction)&ibm64ieee32_ufunc,
};

static char ibm2float32_types[4] = {NPY_UINT32, NPY_FLOAT32, NPY_UINT64,
                                    NPY_FLOAT32};

static void *ibm2float32_data[2] = {NULL, NULL};

PyDoc_STRVAR(ibm2float64_docstring,
             "\
Convert IBM-format single- or double-precision float data, represented\n\
using types np.uint32 or np.uint64 (respectively), to np.float64.\n\
\n\
Examples\n\
--------\n\
>>> ibm2float64(np.uint32(0xc1180000))\n\
-1.5\n\
>>> ibm2float64(np.uint64(0x413243f6a8885a31))\n\
3.141592653589793\n\
>>> ibm2float64(np.uint32(0x61100000))\n\
3.402823669209385e+38\n\
");

PyUFuncGenericFunction ibm2float64_funcs[2] = {
    /* Cast required to suppress compiler warning for NumPy < 1.19.0.
       See https://github.com/numpy/numpy/issues/15252. */
    (PyUFuncGenericFunction)&ibm32ieee64_ufunc,
    (PyUFuncGenericFunction)&ibm64ieee64_ufunc,
};

static char ibm2float64_types[4] = {NPY_UINT32, NPY_FLOAT64, NPY_UINT64,
                                    NPY_FLOAT64};

static void *ibm2float64_data[2] = {NULL, NULL};

static PyMethodDef IBM2IEEEMethods[] = {{NULL, NULL, 0, NULL}};

#if PY_VERSION_HEX >= 0x03000000

static struct PyModuleDef moduledef = {PyModuleDef_HEAD_INIT,
                                       "_ibm2ieee",
                                       NULL,
                                       -1,
                                       IBM2IEEEMethods,
                                       NULL,
                                       NULL,
                                       NULL,
                                       NULL};

PyMODINIT_FUNC
PyInit__ibm2ieee(void)
{
    PyObject *m, *d;
    PyObject *ibm2float32, *ibm2float64;

    m = PyModule_Create(&moduledef);
    if (!m) {
        return NULL;
    }
    d = PyModule_GetDict(m);

    import_array();
    import_umath();

    ibm2float32 = PyUFunc_FromFuncAndData(
        ibm2float32_funcs, ibm2float32_data, ibm2float32_types, 2, 1, 1,
        PyUFunc_None, "ibm2float32", ibm2float32_docstring, 0);
    PyDict_SetItemString(d, "ibm2float32", ibm2float32);
    Py_DECREF(ibm2float32);

    ibm2float64 = PyUFunc_FromFuncAndData(
        ibm2float64_funcs, ibm2float64_data, ibm2float64_types, 2, 1, 1,
        PyUFunc_None, "ibm2float64", ibm2float64_docstring, 0);
    PyDict_SetItemString(d, "ibm2float64", ibm2float64);
    Py_DECREF(ibm2float64);

    return m;
}

#else /* PY_VERSION_HEX >= 0x03000000 */

PyMODINIT_FUNC
init_ibm2ieee(void)
{
    PyObject *m, *d;
    PyObject *ibm2float32, *ibm2float64;

    m = Py_InitModule("_ibm2ieee", IBM2IEEEMethods);
    if (m == NULL) {
        return;
    }
    d = PyModule_GetDict(m);

    import_array();
    import_umath();

    ibm2float32 = PyUFunc_FromFuncAndData(
        ibm2float32_funcs, ibm2float32_data, ibm2float32_types, 2, 1, 1,
        PyUFunc_None, "ibm2float32", ibm2float32_docstring, 0);
    PyDict_SetItemString(d, "ibm2float32", ibm2float32);
    Py_DECREF(ibm2float32);

    ibm2float64 = PyUFunc_FromFuncAndData(
        ibm2float64_funcs, ibm2float64_data, ibm2float64_types, 2, 1, 1,
        PyUFunc_None, "ibm2float64", ibm2float64_docstring, 0);
    PyDict_SetItemString(d, "ibm2float64", ibm2float64);
    Py_DECREF(ibm2float64);
}

#endif /* PY_VERSION_HEX >= 0x03000000 */
